### PR TITLE
feat(server): advertise reasoning-effort ladders on the raw /v1/models list

### PR DESCRIPTION
**Pull request by n3wr1ch**
## Summary

- Grok Build discovers models through the raw OpenAI list shape at `GET /v1/models` (no `client_version`), and its model picker only enables `/effort` (and accepts `--effort`) for entries that advertise the Grok catalog shape: `supports_reasoning_effort` plus `reasoning_efforts` options.
- The raw list previously emitted only `id`, `object`, `created`, and `owned_by`, so routed models appeared in Grok Build without an effort control even when their ladder was already configured and available on the Codex catalog branch. The chat-completions path already honors `reasoning_effort`; model discovery was the missing link.
- Raw-list entries with non-empty effort ladders now advertise their supported levels. Routed entries use provider-configured tiers and defaults, with the canonical fallback order of configured default, `medium`, `high`, then the first tier. Entries with an empty tier list remain unchanged.
- Native entries preserve the pinned upstream ladder and default for each model. In particular, GPT-5.6 Sol and Terra include `ultra`, while GPT-5.6 Luna intentionally ends at `max`, matching OpenAI Codex's canonical `models.json` catalog.
- The Grok Build guide documents how native and routed effort metadata is exposed.

## Verification

- `bun test ./tests/grok-models-effort-list.test.ts ./tests/codex-catalog.test.ts` — 122 passed, 0 failed. The five focused raw-list tests cover routed metadata, empty-tier omission, canonical default fallbacks, and exact Sol/Terra/Luna native ladders and defaults.
- `bun run typecheck` — passed.
- `bun run lint:gui` — passed.
- `bun run test` — 6869 passed, 8 skipped, 0 failed across 479 files.
- `bun run privacy:scan` — passed.
- `cd docs-site && bun install --frozen-lockfile && bun run build` — passed (151 pages built).
- Live provider check: a configured Kimi ladder (`low`, `high`, `max`, default `max`) is returned by `GET /v1/models` in Grok's expected shape. Sending `reasoning_effort` manually already produced distinct upstream reasoning-token counts (13 at `low`, 74 at `high`), confirming the backend path works and only discovery metadata was missing.

## Checklist

- [x] Targets `dev` and keeps the implementation focused.
- [x] Adds regression coverage for changed server/catalog behavior.
- [x] Updates user-facing documentation.
- [x] Reviews security-sensitive behavior and introduces no new auth, secret, dependency, or workflow surface.
- [x] Passes typecheck, lint, full tests, privacy scan, and docs build.

### CODE REVIEW

**Comment by n3wr1ch** [comment]
Thanks for the reviews — all three comments addressed in `ceb5fe75`:

1. **Codex P2 (default fallback):** the raw list now resolves the default with the same canonical order as `applyReasoningLevels` in `src/codex/catalog/effort.ts` (configured default → medium → high → first tier), so both branches advertise the same default when no `modelDefaultReasoningEfforts` is set. Covered by a new test (ladder without a configured default → `medium`; invalid configured default → `high` for a low/high/max ladder).
2. **Codex P2 (native rows):** native entries are now decorated too via `nativeReasoningEfforts` (upstream snapshot ladder, with max+ultra restored for the gpt-5.6 family), re-exported through the `codex/catalog` facade. `ocx-gpt-*` native models get a working /effort control as well. Verified live: `gpt-5.6-sol` advertises low/medium(default)/high/xhigh/max/ultra.
3. **CodeRabbit (docs):** the guide now states the mirrored tiers describe the proxy-configured routed ladder, not native upstream reasoning support, and notes adapters may emulate reasoning or map levels to provider-specific fields.

Full suite green after the changes: `bun run test` 6513 pass / 0 fail, `bun run typecheck` clean.

**Comment by n3wr1ch** [comment]
Addressed the two follow-up CodeRabbit findings in the latest branch:

- **Native ladder docs:** the Grok Build guide now states separately that native GPT-5.6 entries expose their pinned upstream ladder, while routed entries expose proxy-configured provider tiers that adapters may map or emulate.
- **Complete native regression:** the `gpt-5.6-sol` test now asserts the full pinned value list (`low`, `medium`, `high`, `xhigh`, `max`, `ultra`) in addition to the support flag and `medium` default.

The branch is rebased onto current `dev` (`fa5780d5`) and preserves the new upstream `CatalogGatherBusyError` handling in `/v1/models`. Focused tests: 4 pass / 0 fail; typecheck clean.

**Comment by n3wr1ch** [comment]
Addressed the latest CodeRabbit coverage request in `30c7e97f`: added a routed custom model with only `low`/`max` tiers and an invalid configured default, then asserted `low` is selected and is the only option marked default. This directly exercises the final `efforts[0]` branch of the canonical configured → medium → high → first fallback. I used a custom model id so registry wire-map self-healing cannot add a `high` tier to the fixture. Focused suite: 5 pass / 0 fail; typecheck remains clean.

File: src/server/index.ts [hunk 1]
```diff
@@ -459,9 +459,37 @@ export function startServer(port?: number) {
...
+          ...(isDefault ? { default: true } : {}),
+        });
         const data = [
           ...visibleNativeSlugs(config).map(id => ({ id, object: "model", created: 0, owned_by: "openai" })),
-          ...uniqueCatalogModelsForRawPublicList(goOrdered).map(m => ({ id: m.alias ?? `${m.provider}/${m.id}`, object: "model", created: 0, owned_by: m.owned_by ?? m.provider })),
+          ...uniqueCatalogModelsForRawPublicList(goOrdered).map(m => {
+            const efforts = m.reasoningEfforts ?? [];
+            const defaultEffort = m.defaultReasoningEffort && efforts.includes(m.defaultReasoningEffort)
+              ? m.defaultReasoningEffort
+              : efforts[0];
```
[the remark this answers is not in the document]
Comment (Wibias) [reply src/server/index.ts]: [shipping-github] Fixed on head `38da7912`.

The raw `/v1/models` path now uses the same configured → medium → high → first fallback as `applyReasoningLevels`, covered by the routed-default tests in `tests/grok-models-effort-list.test.ts`.

File: src/server/index.ts [hunk 2]
```diff
@@ -459,9 +459,37 @@ export function startServer(port?: number) {
...
+        // fields are ignored by plain OpenAI clients.
+        const grokEffortOption = (value: string, isDefault: boolean) => ({
+          value,
+          label: `${value[0].toUpperCase()}${value.slice(1)} Effort`,
+          ...(isDefault ? { default: true } : {}),
+        });
         const data = [
           ...visibleNativeSlugs(config).map(id => ({ id, object: "model", created: 0, owned_by: "openai" })),
-          ...uniqueCatalogModelsForRawPublicList(goOrdered).map(m => ({ id: m.alias ?? `${m.provider}/${m.id}`, object: "model", created: 0, owned_by: m.owned_by ?? m.provider })),
+          ...uniqueCatalogModelsForRawPublicList(goOrdered).map(m => {
```
[the remark this answers is not in the document]
Comment (Wibias) [reply src/server/index.ts]: [shipping-github] Fixed on head `38da7912`.

Native rows advertise ladders via `nativeReasoningEfforts`, and the follow-up commit also preserves the upstream-pinned default through `nativeDefaultReasoningEffort` (e.g. `gpt-5.6-sol` => `low`), matching the Codex catalog path.

File: tests/grok-models-effort-list.test.ts [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        { value: "low", label: "Low Effort" },
+        { value: "high", label: "High Effort", default: true },
+        { value: "max", label: "Max Effort" },
+      ]);
+      // Native rows advertise the canonical upstream ladder with the same medium-first
+      // default the Codex catalog resolver uses, so /effort works for ocx-gpt-* too.
+      const native = body.data.find(m => m.id === "gpt-5.6-sol");
+      expect(native).toBeDefined();
+      expect(native!.supports_reasoning_effort).toBe(true);
+      expect(native!.reasoning_effort).toBe("medium");
```
[the remark this answers is not in the document]
Comment (Wibias) [reply tests/grok-models-effort-list.test.ts]: [shipping-github] Already covered on head `38da7912`.

The native `gpt-5.6-sol` assertion checks the full pinned value list `low/medium/high/xhigh/max/ultra` in addition to the support flag and default.

**Comment by Wibias** [comment]
## [shipping-github] Verdict: changes-requested

**PR:** `#853` — feat(server): advertise reasoning-effort ladders on the raw `/v1/models` list
**Head:** `c0426f67` on `dev` (mergeStateStatus: `UNSTABLE`, mergeable)
**Linked:** none

### Usefulness

Useful and in scope: it makes configured routed reasoning ladders, and pinned native ladders, discoverable by Grok Build through the plain OpenAI `GET /v1/models` surface so `/effort` can be enabled consistently.

### Bugs / correctness

- Complementary deep review completed for silent failures, resource leaks, edge cases, API compatibility, resource lifecycle, boundary conditions, filesystem/test cleanup, error propagation, and network cancellation. No open blocking finding remains.
- A native-default mismatch found during this review was fixed in `38da7912`: `gpt-5.6-sol` now preserves the upstream `low` default through `nativeDefaultReasoningEffort`, with a regression assertion in `tests/grok-models-effort-list.test.ts`.

### Security

- No confirmed security finding. The change only serializes server-controlled catalog metadata into JSON; it does not add authentication, authorization, outbound destinations, shell/HTML evaluation, secrets, or dependency/workflow changes.
- AI/agent/MCP review completed: this is model metadata discovery, not tool invocation, prompt construction, MCP registration, or agent-permission expansion.

### Spec / standards

- Spec source: PR body and existing catalog behavior. The implementation covers routed/native ladders, empty-tier omission, canonical defaults, native upstream defaults, focused tests, and docs.
- The PR body’s verification paragraph is stale (`3` new tests / `6512` pass and the old first-tier-only fallback description); the current branch has five focused tests and the canonical medium/high/first fallback plus native default preservation. Please refresh that text.

### Reviews

- Owners/maintainers: no open human threads.
- Bots: all six review threads are resolved; CodeRabbit and Codex findings were verified and addressed or superseded by later commits.

### Base / CI

- Updated from current `upstream/dev` in `c0426f67`; no conflicts and the effective PR diff remains five focused files.
- Local `bun test tests/grok-models-effort-list.test.ts` passed 5/5, `bun run typecheck` passed, and `bun run privacy:scan` passed. The full `bun run test` attempt exceeded the 10-minute local command budget without completing.
- Required workflows for the new fork head are not green: `Cross-platform CI` and `React Doctor` are `action_required` with no jobs, so GitHub has not executed them on `c0426f67`. The prior head `38da7912` had green cross-platform and React Doctor runs.

### Gate

Hard blocker: approve/execute the two `action_required` fork workflows, then wait for their results on `c0426f67`. This is a GitHub permission/approval state, not a reported code failure.

### Bottom line

The implementation is useful, focused, locally validated, and has no remaining confirmed bug or security blocker. I cannot give an `approve-comment` verdict while the required workflows have not run on the current synced head; after the fork workflow approval completes and the PR description counts are refreshed, rerun the final gate.

**Comment by lidge-jun** [comment]
Pre-merge review result: **not merge-ready yet**. @Wibias's verification covered the native-default fix, but one independently reproduced mismatch remains:

1. `src/codex/catalog/metadata.ts:95` adds `ultra` to every GPT-5.6 native, so raw `/v1/models` advertises Luna as `low…ultra` while the canonical catalog intentionally ends Luna at `max` (`src/codex/catalog/sync.ts:145`, `tests/codex-catalog.test.ts:1113`). Probe: raw helper returns `[low, medium, high, xhigh, max, ultra]` vs canonical `[low, medium, high, xhigh, max]`. This would expose a noncanonical effort option to Grok Build.
2. `tests/grok-models-effort-list.test.ts:61` only asserts Sol's ladder (which legitimately contains `ultra`), so it cannot catch the Luna discrepancy — please add Luna/Terra assertions.
3. The `windows` CI job was still in progress at review time; full green CI is required before merge. The PR body is also stale (says three focused tests; head has five).

Otherwise the shape is good: the `/v1/models` change is additive-only and the default-ladder fallback order matches canonical behavior. Small fix, fast re-review.

**Comment by n3wr1ch** [comment]
Addressed the remaining native-ladder mismatch in `53d83740`.

- `nativeReasoningEfforts` now preserves the exact pinned per-model ladder instead of adding `ultra` to every GPT-5.6 native.
- Raw `/v1/models` now advertises Sol and Terra through `ultra`, while Luna correctly ends at `max`; defaults remain Sol=`low`, Terra/Luna=`medium`.
- The focused integration test now asserts all three complete ladders and defaults.
- Refreshed the PR description with current behavior and verification counts.

Verification: focused catalog/server tests 122/122, typecheck, GUI lint, full suite 6869 pass / 8 skip / 0 fail, privacy scan, and docs build (151 pages) all passed.

**Comment by n3wr1ch** [comment]
The latest `windows` job has one unrelated timeout in `tests/codex-auth-context.test.ts`:

```text
Codex auth context > busy and stale Codex refresh failures do not mark the auth-context account for reauth
This test timed out after 5000ms.
```

The run otherwise completed with 6873 passing tests, and the PR's `/v1/models` / GPT-5.6 ladder tests passed. Ubuntu, macOS, npm-global Windows, typecheck, and the local full suite are green. This auth-context test and its runtime code are outside the five-file PR diff, so this appears to be a Windows timing flake rather than a regression from this change.

Could a maintainer please rerun the failed Windows job?

**Comment by Wibias** [comment]
[shipping-github] Addressed feedback
feedback: issue_comment:5154105797
commit: 53d837405d1d2736034264e4ac85012a97090d8a

Verified on `dd813da2`: Luna's raw `/v1/models` ladder ends at `max`, Sol and Terra retain `ultra`, and the focused catalog/server tests pass 118/118. The branch is also updated through `dev@6ac6beaf`.

**Comment by Wibias** [comment]
Thanks @n3wr1ch — merging this.

Why it helps: raw `GET /v1/models` now advertises the exact reasoning-effort ladders Grok Build needs, so routed and native models expose `/effort` without widening Luna beyond its canonical `max`. Sol and Terra retain `ultra`, defaults follow the canonical fallback, and complete ladder regressions cover the public response.

Ship it.
